rust-bins/src/bin/keygen.rs

```diff
@@ -301,6 +472,228 @@ pub fn generate_ed_sk(
     Ok(sk)
 }
 
+/// Asks user to input word with given number and reads it from stdin.
+pub fn read_word(number: u8) -> Result<String, std::io::Error> {
+    print!("Word {}: ", number);
+    io::stdout().flush()?; // actually print before reading
+    let mut line = String::new();
+    io::stdin().read_line(&mut line)?;
+    let word = line.trim().to_string(); // remove trailing newline
+    Ok(word)
+}
+
+/// Asks user to input word with given number and reads it from stdin.
+/// Then checks whether the word is in valid_words and starts over if not.
+pub fn read_bip39_word(
+    number: u8,
+    bip39_map: &HashMap<&str, usize>,
+) -> Result<String, std::io::Error> {
+    loop {
+        let word = read_word(number)?;
+        if bip39_map.contains_key(&*word) {
+            return Ok(word);
+        } else {
+            println!("The word you have entered is not in the BIP39 word list. Please try again.");
+        }
+    }
+}
+
+/// Read vector of words from file.
+pub fn read_words_from_file(
+    path: PathBuf,
+    verify_bip39: bool,
+    bip39_map: &HashMap<&str, usize>,
+) -> Result<Vec<String>, String> {
+    let word_string = succeed_or_die!(
+        fs::read_to_string(path),
+        e => "Could not read input from provided file because {}"
+    );
```

Comment:
While it might not be too terrible in this case, I think it is bad practice to read entire  files into memory when you don't know their size.

Maybe it would be better to do something along the lines of
```rust
let file = succeed_or_die!(fs::File::open(path), ...);
let reader = BufReader::new(file);
let word_list: Vec<String> = reader.lines().flat_map(|x|  x.split_whitespace().map(str::to_owned)).collect();
```

Reply:
Doesn't your code also read the entire file? Maybe one should stop after reading 24 lines. But what if the first line is very long?

Reply:
It stores its content only once though and in this case we need to store all the words so it must read those.

It doesn't matter so much in this case.